File: app/schemas/wellness.py
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, Field

from app.models.wellness import MoodLevel, EnergyLevel


class WellnessCheckInCreate(BaseModel):
    mood: MoodLevel
    energy: EnergyLevel
    stress_level: int = Field(..., ge=1, le=5, description="Stress level from 1 (low) to 5 (high)")
    sleep_hours: Optional[float] = Field(None, ge=0, le=24, description="Hours of sleep last night")
    notes: Optional[str] = Field(None, max_length=1000)
    is_private: bool = False


class WellnessCheckInResponse(BaseModel):
    id: int
    student_id: int
    mood: MoodLevel
    energy: EnergyLevel
    stress_level: int
    sleep_hours: Optional[float] = None
    notes: Optional[str] = None
    is_private: bool
    check_in_date: date
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class DayTrendPoint(BaseModel):
    date: date
    mood: Optional[MoodLevel] = None
    energy: Optional[EnergyLevel] = None
    stress_level: Optional[int] = None
    sleep_hours: Optional[float] = None
    has_entry: bool = False


class WellnessTrendResponse(BaseModel):
    days: list[DayTrendPoint]
    avg_stress: Optional[float] = None
    avg_sleep: Optional[float] = None
    dominant_mood: Optional[MoodLevel] = None
    dominant_energy: Optional[EnergyLevel] = None
    streak_days: int = 0  # consecutive days checked in


class WellnessSummary(BaseModel):
    student_id: int
    week_avg_stress: Optional[float] = None
    week_avg_sleep: Optional[float] = None
    dominant_mood: Optional[MoodLevel] = None
    dominant_energy: Optional[EnergyLevel] = None
    alert_active: bool = False
    total_check_ins_this_week: int = 0
    streak_days: int = 0

File: frontend/src/api/wellness.ts
import { api } from './client';

// -------------------------------------------------------------------------
// Enums (mirror Python enums)
// -------------------------------------------------------------------------

export type MoodLevel = 'great' | 'good' | 'okay' | 'struggling' | 'overwhelmed';
export type EnergyLevel = 'high' | 'medium' | 'low';

// -------------------------------------------------------------------------
// Request / response types
// -------------------------------------------------------------------------

export interface WellnessCheckInCreate {
  mood: MoodLevel;
  energy: EnergyLevel;
  stress_level: number; // 1-5
  sleep_hours?: number | null;
  notes?: string | null;
  is_private?: boolean;
}

export interface WellnessCheckInResponse {
  id: number;
  student_id: number;
  mood: MoodLevel;
  energy: EnergyLevel;
  stress_level: number;
  sleep_hours: number | null;
  notes: string | null;
  is_private: boolean;
  check_in_date: string; // ISO date "YYYY-MM-DD"
  created_at: string;
  updated_at: string | null;
}

export interface DayTrendPoint {
  date: string; // "YYYY-MM-DD"
  mood: MoodLevel | null;
  energy: EnergyLevel | null;
  stress_level: number | null;
  sleep_hours: number | null;
  has_entry: boolean;
}

export interface WellnessTrendResponse {
  days: DayTrendPoint[];
  avg_stress: number | null;
  avg_sleep: number | null;
  dominant_mood: MoodLevel | null;
  dominant_energy: EnergyLevel | null;
  streak_days: number;
}

export interface WellnessSummary {
  student_id: number;
  week_avg_stress: number | null;
  week_avg_sleep: number | null;
  dominant_mood: MoodLevel | null;
  dominant_energy: EnergyLevel | null;
  alert_active: boolean;
  total_check_ins_this_week: number;
  streak_days: number;
}

// -------------------------------------------------------------------------
// API calls
// -------------------------------------------------------------------------

export const wellnessApi = {
  /** Create or update today's wellness check-in (student only). */
  submitCheckIn(data: WellnessCheckInCreate): Promise<WellnessCheckInResponse> {
    return api.post<WellnessCheckInResponse>('/api/wellness/check-in', data).then(r => r.data);
  },

  /** Get today's check-in for the authenticated student, or null. */
  getToday(): Promise<WellnessCheckInResponse | null> {
    return api.get<WellnessCheckInResponse | null>('/api/wellness/today').then(r => r.data);
  },

  /** Get trend for the authenticated student. */
  getTrend(days: number = 7): Promise<WellnessTrendResponse> {
    return api.get<WellnessTrendResponse>('/api/wellness/trend', { params: { days } }).then(r => r.data);
  },

  /** Get non-private summary for a student (teacher/admin/parent). */
  getStudentSummary(studentId: number): Promise<WellnessSummary> {
    return api
      .get<WellnessSummary>(`/api/wellness/student/${studentId}/summary`)
      .then(r => r.data);
  },

  /** Get non-private trend for a child (parent only). */
  getChildTrend(studentId: number, days: number = 7): Promise<WellnessTrendResponse> {
    return api
      .get<WellnessTrendResponse>(`/api/wellness/student/${studentId}/trend`, { params: { days } })
      .then(r => r.data);
  },
};
